app/src/main/java/mozilla/lockbox/presenter/AutofillRoutePresenter.kt

```diff
@@ -101,51 +121,61 @@ class AutofillRoutePresenter(
             log.error(
                 "Cannot route from $from to $to. " +
-                    "This is a developer bug, fixable by adding an action to graph_autofill.xml"
+                    "This is a developer bug, fixable by adding an action to $graphName.xml and/or ${javaClass.simpleName}"
             )
+            null
         } else {
-            val clearBackStack = src.getAction(transition)?.navOptions?.shouldLaunchSingleTop() ?: false
-            if (clearBackStack) {
-                while (navController.popBackStack()) {
-                    // NOP
+            // Get the transition action out of the graph, before we manually clear the back
+            // stack, because it causes IllegalArgumentExceptions.
+            src.getAction(transition)?.navOptions?.let { navOptions ->
+                if (navOptions.shouldLaunchSingleTop()) {
+                    while (navController.popBackStack()) {
+                        // NOP
+                    }
+                    routeStore.clearBackStack()
                 }
+                navOptions
             }
         }
 
         try {
-            navController.navigate(transition, args)
+            navController.navigate(destinationId, args, navOptions)
         } catch (e: IllegalArgumentException) {
             log.error("This appears to be a bug in navController", e)
             navController.navigate(destinationId, args)
         }
     }
 
-    private fun findTransitionId(@IdRes from: Int, @IdRes to: Int): Int? {
-        return when (Pair(from, to)) {
-            Pair(R.id.fragment_locked, R.id.fragment_filter) -> R.id.action_locked_to_filter
-            Pair(R.id.fragment_null, R.id.fragment_filter) -> R.id.action_to_filter
-            Pair(R.id.fragment_null, R.id.fragment_locked) -> R.id.action_to_locked
+    private fun findTransitionId(@IdRes from: Int, @IdRes destination: Int): Int? {
+        return when (from to destination) {
+            R.id.fragment_locked to R.id.fragment_filter_backdrop -> R.id.action_to_filter
+            R.id.fragment_filter_backdrop to R.id.fragment_locked -> R.id.action_to_locked
             else -> null
         }
     }
 
     private fun showDialogFragment(dialogFragment: DialogFragment, destination: RouteAction.DialogFragment) {
-        val fragmentManager = activity.supportFragmentManager
         try {
-            dialogFragment.show(fragmentManager, dialogFragment.javaClass.name)
+            dialogFragment.setTargetFragment(currentFragment, 0)
+            dialogFragment.show(navHostFragmentManager, dialogFragment.javaClass.name)
             dialogFragment.setupDialog(destination.dialogTitle, destination.dialogSubtitle)
         } catch (e: IllegalStateException) {
             log.error("Could not show dialog", e)
         }
     }
 
+    private fun dismissDialogIfPresent() {
+        (currentFragment as? DialogFragment)?.dismiss()
+    }
```

Comment:
✅ ✅ ✅

Reply:
Why isn't the `dialog` set up to be dismissed on its own? Should we be doing this same thing for the `RoutePresenter`'s `showDialogFragment`? https://github.com/mozilla-lockbox/lockbox-android/blob/ac8add36bd4c79ad0b6db11881b5da928a2bbaa2/app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt#L166

Reply:
We don't have the same situation in the normal app; we rarely have to dismiss a dialog _without_ anything being tapped on it.